### PR TITLE
Add a navigation mode for message scroll

### PR DESCRIPTION
**Pull request by cobwebsonsale**
- Disable scroll to bottom on new messages if user
  scrolls up to read messages

### CODE REVIEW

**Comment by cobwebsonsale** [comment]
here's how the behaviour looks like:
nav_mode

**Comment by olavfosse** [comment]
Will be reviewing this now.

File: front/src/components/Messages.js [hunk 1]
```diff
@@ -55,14 +55,21 @@ const MessagesDiv = styled.div`
 
 
 const Messages = (props) => {
+  const [isNavigationMode, setNavigationMode] = useState()
```
Comment (olavfosse) [inline front/src/components/Messages.js]: `const [isNavigationMode, setIsNavigationMode] = useState(false)`
is better

Comment (cobwebsonsale) [reply front/src/components/Messages.js]: Can I push another commit for this edit or should I amend my last commit? How do you prefer things?

**Comment by olavfosse** [comment]
Exiting navigation-mode, by scrolling all the way down, did not work on my box because `element.scrollTop + element.offSetHeight` and `element.scrollHeight` are slightly different even though I've scrolled all the way down I can. I'll fix this by changing the condition to only require that they are very close to each other. I'm thinking `3px` should be sufficient.

---
```
$ uname -a
Linux Shiina 5.4.0-52-generic #57-Ubuntu SMP Thu Oct 15 10:57:00 UTC 2020 x86_64 x86_64 x86_64 GNU/Linux
```

```
$ chromium --version
Chromium 86.0.4240.111 snap
```

File: front/src/components/Messages.js (hunk 1 above)
Comment (olavfosse) [reply front/src/components/Messages.js]: push another. It's best to avoid changing public history.

File: front/src/components/Messages.js [hunk 2]
```diff
@@ -55,14 +55,21 @@ const MessagesDiv = styled.div`
...
   const messagesEndRef = useRef(null)
 
   useEffect(() => {
-    messagesEndRef.current.scrollIntoView({ behavior: "smooth" })
+    !isNavigationMode && messagesEndRef.current.scrollIntoView({ behavior: "smooth" })
   }, [props.messages])
 
+  const handleScroll = (scrollEvent) => {
+      var element = scrollEvent.currentTarget
+      var isBottom = element.scrollTop + element.offsetHeight === element.scrollHeight
```
Comment (olavfosse) [inline front/src/components/Messages.js]: avoid `var`. Use `const` for variables that you don't mutate and `let` for those that you do.

File: front/src/components/Messages.js [hunk 3]
```diff
@@ -55,14 +55,21 @@ const MessagesDiv = styled.div`
...
+      isBottom ? setNavigationMode(false) : setNavigationMode(true)
```
Comment (olavfosse) [inline front/src/components/Messages.js]: This is logically equivalent to
`setNavigationMode(isBottom)`

**Comment by cobwebsonsale** [comment]
>  I'll fix this by changing the condition to only require that they are very close to each other. I'm thinking 3px should be sufficient.

Makes sense. Stackoverflow had similar suggestions around testing the difference of two values to be less than 1 unit. (unsure what the unit was)

File: front/src/components/Messages.js (hunk 1 above)
Comment (cobwebsonsale) [reply front/src/components/Messages.js]: Cool. This is done.

File: front/src/components/Messages.js (hunk 3 above)
Comment (cobwebsonsale) [reply front/src/components/Messages.js]: oh you're right. let me fix this.

Comment (olavfosse) [reply front/src/components/Messages.js]: Sorry, it should of course be `setNavigationMode(!isBottom)`.

Comment (cobwebsonsale) [reply front/src/components/Messages.js]: Done. Changed `isBottom` to `!isBottom` though.

**Comment by cobwebsonsale** [comment]
I've added the error margin for reaching bottom. Can you please check if it's enough when you scroll?

**Comment by olavfosse** [comment]
Works great here!

**Comment by olavfosse** [comment]
I am just adding some comments and small refinements then I will merge.

**Comment by olav35** [comment]
What browser/os did you make the gif with?

**Comment by cobwebsonsale** [comment]
Safari/Mac
``` 
$ uname -a
Darwin 19.6.0 Darwin Kernel Version 19.6.0: Thu Jun 18 20:49:00 PDT 2020; root:xnu-6153.141.1~1/RELEASE_X86_64 x86_64
```
```
Safari
Version 14.0 (15610.1.28.1.9, 15610)
```

The browser actually scrolls _beyond_ the height, where the difference becomes negative.

**Comment by olavfosse** [comment]
Merging now. Thanks!

**Pull request merged**